src/lib/io/include/sol-uart.h

```diff
@@ -238,7 +251,18 @@ void sol_uart_close(struct sol_uart *uart);
  * @note Caller should guarantee that tx buffer will not be freed until
  * callback is called.
  */
-bool sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
+int sol_uart_write(struct sol_uart *uart, const uint8_t *tx, size_t length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
```

Comment:
see PR #1910 with improved behavior. The goal is to find a common pattern for all functions that may have pending information and will asynchronously deliver progressive data to you, so HTTP, UART, socket...

In order to save memory on small systems, instead of keeping a buffer that is realloc()ed to fit pending memory, we take memory segments (blobs) that can be reference counted and have hierarchy (so one can have windows/slices, take references to that and it keeps references to parent). It's a bit more cumbersome to use, but at least making it a common pattern that is efficient on small OS helps. Usability we can help with wrappers.

This also address you request: which data was written/failed? You can use the blob to identify that.

As for one-callback, this is the standard use case where you have only one callback per instance. Thus makes it easier to use and consumes less memory.